### ISSUE TITLE
Problem creating dlls for mIRC (the problem can be bigger)

### ISSUE THREAD

**Issue by rockcavera** [brechtsanders/winlibs_mingw#87]
I tested your gcc and clang compilers (MSVCRT i386) to compile dlls for mIRC. The problem occurs when I try to use the dlls in mIRC as I get this error: "* /dll: unable to open file 'D:\mIRC\test.dll'".

The strange thing is that with the clang distributed here https://github.com/mstorsjo/llvm-mingw and https://llvm.org/, as well as with gcc from tdm64 and gcc from [mingw64](https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/ray_linn/GCC-10.x-with-ada/), it works correctly. Which leads me to believe there is a problem in your compilation of the compilers.

I'll do more testing to see if the problem extends to non-mIRC usage dlls.

Tested releases and snapshots:
```
winlibs-i686-posix-dwarf-gcc-11.2.0-llvm-13.0.0-mingw-w64-9.0.0-r3
winlibs-i686-posix-dwarf-gcc-10.3.0-llvm-12.0.0-mingw-w64-9.0.0-r2
winlibs-i686-posix-dwarf-gcc-11.2.0-mingw-w64-9.0.0-r5
winlibs-i686-posix-dwarf-gcc-11.2.0-mingw-w64-9.0.0-r3
winlibs-i686-posix-dwarf-gcc-10.3.0-mingw-w64-9.0.0-r2
winlibs-i686-posix-dwarf-gcc-9.4.0-mingw-w64-9.0.0-r2
winlibs-i686-posix-dwarf-gcc-8.5.0-mingw-w64-9.0.0-r1
winlibs-i686-posix-dwarf-gcc-7.5.0-mingw-w64-7.0.0-r1
winlibs-i686-posix-dwarf-gcc-12.0.1-snapshot20220123-mingw-w64-9.0.0-r1
winlibs-i686-posix-dwarf-gcc-11.2.1-snapshot20211204-mingw-w64-9.0.0-r1
```

**Comment by brechtsanders** [comment]
Can you point me to the source code you are trying to build and can you post more output of the failing build process?

It might also be a good idea to run a tool like [DependencyWalker](https://www.dependencywalker.com/) on D:\mIRC\test.dll to check what it depends on and if those DLLs are available in the PATH in the environment where you get the error.

**Comment by rockcavera** [comment]
I found the cause of the problem thanks to Dependency Walker. The problem is that your compiler dynamically links libgcc and generating a dependency on the dll libgcc_s_dw2-1.dll which in turn depends on libwinpthread-1.dll. I compiled it with -static-libgcc and it solved the problem.

Just one question before closing the issue. Why is libgcc linked dynamically and not statically by default?

**Comment by rockcavera** [comment]
File test.c
```c
#include <windows.h>
#include <stdio.h>

#ifdef  __cplusplus
#  define MIRC_PROC(rettype, name) extern "C" rettype __stdcall name
#else
#  define MIRC_PROC(rettype, name) rettype __stdcall name
#endif

#if defined(_MSC_VER)
#  pragma comment(linker, "/export:LoadDll=_LoadDll@4")
#  pragma comment(linker, "/export:UnloadDll=_UnloadDll@4")
#  pragma comment(linker, "/export:test=_test@24")
#endif

DWORD major;
DWORD minor;

typedef struct {
  DWORD mVersion;
  HWND mHwnd;
  BOOL mKeep;
  BOOL mUnicode;
  DWORD mBeta;
  DWORD mBytes;
} LOADINFO;

MIRC_PROC(void, LoadDll)(LOADINFO *loadInfo) {
  loadInfo->mKeep = FALSE;
  major = loadInfo->mVersion & 0xFFFFUL;
  minor = (loadInfo->mVersion >> 16) & 0xFFFFUL;
}

MIRC_PROC(int, UnloadDll)(int mTimeout) {
  return 1;
}

MIRC_PROC(int, test)(HWND mWnd, HWND aWnd, char *data, char *parms, BOOL show, BOOL nopause) {
  sprintf(data, "/echo -a Major: %lu Minor: %lu Dif:", major, minor);
  return 2;
}
```
File test.def
```
EXPORTS
LoadDll = LoadDll@4
UnloadDll = UnloadDll@4
test = test@24
```

Compiled with doesn't work for those who don't have libgcc_s_dw2-1.dll: `gcc -shared -O3 test.def -o test.dll test.c`

Compiled with already works: `gcc -shared -static-libgcc -O3 test.def -o test.dll test.c`

I'm not sure, but your compiler seems to have an alias to dynamically link libgcc when passed -shared. In my opinion it should be -shared-libgcc.

**Comment by brechtsanders** [comment]
The idea of shared libraries (DLL files on Windows) is that common code is bundled in separate files that can be called from multiple executable files (EXE files on Windows) or other shared libraries.

So if you want to distribute shared builds you will always need to make sure the dependency libraries are also distributed.

This is true for Windows as it is on other operating systems, but on Linux distributions you would typically have a package manager pulling in the dependencies during installation.
On Windows you should copy the `.dll` files in the same folder as the `.exe` file(s) (or make sure they can be found via the `PATH` environment variable).

But if you don't want that, you should build static libraries using flags like `-static` and `-static-libgcc` (and `-static-libstdc++` for C++), so the `.exe` files don't need any `.dll` files to run (apart from any system `.dll` files it may use).

In my [pedeps](https://github.com/brechtsanders/pedeps/) I included a tool `copypedeps` which can be used with the `-r` flag to copy `.exe`/`.dll` files along with all dependencies allowing you to bundle your application's distribution package.

In conclusion: I will close this ticket, as what you mention is how shared libraries work (and MinGW-w64 specifically with dependencies on its standard and winpthread libraries). If you find the whole DLL thing too complex - as many do, that's why some call it DLL hell - you should stick with Windows.

P.S.: It may be interesting to you that it is possible on Windows to build a DLL that is static with respect to it's dependencies (= it doesn't depend on any other DLL files).
